src/test/regress/expected/multi_explain.out

```diff
@@ -2286,15 +2286,14 @@ Custom Scan (Citus Adaptive) (actual rows=1 loops=1)
 EXPLAIN :default_analyze_flags
 SELECT count(distinct a) FROM (SELECT GREATEST(random(), 2) r, a FROM dist_table) t NATURAL JOIN ref_table;
 Aggregate (actual rows=1 loops=1)
-  ->  Custom Scan (Citus Adaptive) (actual rows=10 loops=1)
+  ->  Custom Scan (Citus Adaptive) (actual rows=4 loops=1)
```

Comment:
this confused me a bit at first, but it's actually the most significant performance benefit of this change: Count distinct on the distribution column now gets pushed down even when there are subqueries.

Reply:
One interesting observation I have is that sometimes the old implementation performs even faster. I initially thought the performance difference is because of `random()` is interpreted differently in the query, but it seems unrelated? 

Common:
```SQL
insert into users_table(user_id, value_1) SELECT i%100,i%100 from generate_series(0,100000)i;
```


Branch:
```SQL
 explain (analyze, verbose) select count(DISTINCT u1_id) FROM (SELECT u1.user_id as u1_id, random() FROM users_table u1 JOIN users_table u2 USING(user_id)) as foo;

                                                                                                                                                                                        QUERY PLAN                                                                                                                                                                                         
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=250.00..250.02 rows=1 width=8) (actual time=9886.651..9886.651 rows=1 loops=1)
   Output: COALESCE((pg_catalog.sum(remote_scan.count))::bigint, '0'::bigint)
   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8) (actual time=9886.576..9886.589 rows=32 loops=1)
         Output: remote_scan.count
         Task Count: 32
         Tuple data received from nodes: 32 bytes
         Tasks Shown: One of 32
         ->  Task
               Query: SELECT count(DISTINCT worker_column_1) AS count FROM (SELECT foo.u1_id AS worker_column_1 FROM (SELECT u1.user_id AS u1_id, random() AS random FROM (public.users_table_102176 u1(user_id, "time", value_1, value_2, value_3, value_4) JOIN public.users_table_102176 u2(user_id, "time", value_1, value_2, value_3, value_4) USING (user_id))) foo) worker_subquery
               Tuple data received from node: 1 bytes
               Node: host=localhost port=9700 dbname=postgres
               ->  Aggregate  (cost=51617.60..51617.61 rows=1 width=8) (actual time=7987.596..7987.596 rows=1 loops=1)
                     Output: count(DISTINCT u1.user_id)
                     ->  Hash Join  (cost=104.15..27121.68 rows=1959674 width=12) (actual time=4.047..3340.513 rows=6000000 loops=1)
                           Output: u1.user_id, random()
                           Hash Cond: (u1.user_id = u2.user_id)
                           ->  Seq Scan on public.users_table_102176 u1  (cost=0.00..61.29 rows=3429 width=4) (actual time=0.045..2.205 rows=6000 loops=1)
                                 Output: u1.user_id, u1."time", u1.value_1, u1.value_2, u1.value_3, u1.value_4
                           ->  Hash  (cost=61.29..61.29 rows=3429 width=4) (actual time=3.942..3.943 rows=6000 loops=1)
                                 Output: u2.user_id
                                 Buckets: 8192 (originally 4096)  Batches: 1 (originally 1)  Memory Usage: 275kB
                                 ->  Seq Scan on public.users_table_102176 u2  (cost=0.00..61.29 rows=3429 width=4) (actual time=0.012..2.267 rows=6000 loops=1)
                                       Output: u2.user_id
                   Planning Time: 0.619 ms
                   Execution Time: 7988.100 ms
 Planning Time: 2.213 ms
 Execution Time: 9887.573 ms
(27 rows)

Time: 9899.024 ms (00:09.899)
```

Master:
```SQL
 Aggregate  (cost=250.00..250.01 rows=1 width=8) (actual time=976.020..976.020 rows=1 loops=1)
   Output: count(DISTINCT remote_scan.count)
   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=4) (actual time=975.943..975.944 rows=3 loops=1)
         Output: remote_scan.count
         Task Count: 1
         Tuple data received from nodes: 5 bytes
         Tasks Shown: All
         ->  Task
               Query: SELECT worker_column_1 AS count FROM (SELECT foo.u1_id AS worker_column_1 FROM (SELECT u1.user_id AS u1_id, random() AS random FROM (public.users_table_102168 u1(user_id, "time", value_1, value_2, value_3, value_4) JOIN public.users_table_102168 u2(user_id, "time", value_1, value_2, value_3, value_4) USING (user_id))) foo) worker_subquery GROUP BY worker_column_1
               Tuple data received from node: 5 bytes
               Node: host=localhost port=9700 dbname=postgres
               ->  HashAggregate  (cost=78883.00..78883.03 rows=3 width=4) (actual time=907.161..907.161 rows=3 loops=1)
                     Output: u1.user_id
                     Group Key: u1.user_id
                     ->  Hash Join  (cost=81.50..41383.00 rows=3000000 width=12) (actual time=1.135..423.955 rows=3000000 loops=1)
                           Output: u1.user_id, random()
                           Hash Cond: (u1.user_id = u2.user_id)
                           ->  Seq Scan on public.users_table_102168 u1  (cost=0.00..44.00 rows=3000 width=4) (actual time=0.017..0.934 rows=3000 loops=1)
                                 Output: u1.user_id, u1."time", u1.value_1, u1.value_2, u1.value_3, u1.value_4
                           ->  Hash  (cost=44.00..44.00 rows=3000 width=4) (actual time=1.034..1.034 rows=3000 loops=1)
                                 Output: u2.user_id
                                 Buckets: 4096  Batches: 1  Memory Usage: 138kB
                                 ->  Seq Scan on public.users_table_102168 u2  (cost=0.00..44.00 rows=3000 width=4) (actual time=0.008..0.523 rows=3000 loops=1)
                                       Output: u2.user_id
                   Planning Time: 3.256 ms
                   Execution Time: 907.256 ms
 Planning Time: 26.596 ms
 Execution Time: 976.301 ms
(28 rows)

Time: 1062.255 ms (00:01.062)
```